### ISSUE TITLE
[2.0.18] duplicate default routes in kernel routing table

### ISSUE THREAD

**Issue by ahmetalpbalkan** [Azure/WALinuxAgent#393]
Verbatim copy from an internal incident report:

```
Problem: output for route command , duplicate default gateways.

~]$ route
Kernel IP routing table
Destination Gateway Genmask Flags Metric Ref Use Iface
default 10.0.0.1 0.0.0.0 UG 0 0 0 eth0
default 10.0.0.1 0.0.0.0 UG 0 0 0 eth0
10.0.0.0 0.0.0.0 255.255.255.0 U 0 0 0 eth0
168.63.129.16 10.0.0.1 255.255.255.255 UGH 0 0 0 eth0
link-local 0.0.0.0 255.255.0.0 U 1002 0 0 eth0


Additional Notes:
I recreated the issue with multiple default gateway in different Linux Distors
and I almost convinced that this is a waagent version. 

First I created a Centos7.2, with waagent 2.018, installed by default.

=========== 
[root@centos72 ~]# rpm -qa | grep WALinuxAgent
WALinuxAgent-2.0.18-1.noarch
=============

I removed the installed rpm “WALinuxAgent-2.0.18-1.noarch”, and downloaded
a lower version “WALinuxAgent-2.0.16-1.el7.noarch.rpm”. I installed that version
and restarted the network script in the VM.
After that I could see that default gateway is only one instead of two.

============ 
[root@centos72 ~]# route
Kernel IP routing table
Destination Gateway Genmask Flags Metric Ref Use Iface
default 10.0.0.1 0.0.0.0 UG 0 0 0 eth0
10.0.0.0 0.0.0.0 255.255.255.0 U 0 0 0 eth0
168.63.129.16 10.0.0.1 255.255.255.255 UGH 0 0 0 eth0
link-local 0.0.0.0 255.255.0.0 U 1002 0 0 eth0
=============== 

So I saved a copy of the /usr/sbin/waagent, which is created from
“WALinuxAgent-2.0.16-1.el7.noarch.rpm” as a file by the name, /root/waagent.bak.
Then I upgraded the waagent back to the version “WALinuxAgent-2.0.18-1.noarch”. 

After that I compared the files /usr/sbin/waagent(from WALinuxAgent-2.0.18-1.noarch)
and backup file /root/waagent.bak(WALinuxAgent-2.0.16-1.el7.noarch.rpm).

I could see the following differences in the files, corresponding to the line numbers.
================= 
759 def setDefaultGateway(self, gateway): def setDefaultGateway(self, gateway):
760 addRouteResult = Run("/sbin/route add default gw " + | Run("/sbin/route add default gw" + gateway, chk_err=F
761 Log("add default gateway result: {0}".format(addRoute <
762
763 def routeAdd(self, net, mask, gateway): def routeAdd(self, net, mask, gateway):
764 Run("/sbin/route add -net " + net + " netmask " + mas Run("/sbin/route add -net " + net + " netmask " + mas
765 chk_err=False) chk_err=False)
================================================ 
1701 def setDefaultGateway(self, gateway): <
1702 addRouteResult = Run("route add default gw " + gatewa <
1703 Log("add default gateway result: {0}".format(addRoute <
==============================
2887 def setDefaultGateway(self, gateway): def setDefaultGateway(self, gateway):
2888 Run("/sbin/route add default " + gateway, chk_err=Fal Run("/sbin/route add default " + gateway, chk_err=Fal
==================
6050 def SetDefaultGateway(self, gateway): def SetDefaultGateway(self, gateway):
6051 """ """
6052 Set default gateway Set default gateway
6053 """ """
6054 gateway = self.IntegerToIpAddressV4String(gateway) gateway = self.IntegerToIpAddressV4String(gateway)
6055 Log("Set default gateway: {0}".format(gateway)) Log("Set default gateway: {0}".format(gateway))
6056 MyDistro.setDefaultGateway(gateway) MyDistro.setDefaultGateway(gateway)
=====================
6138 if option == 3: if option == 3:
6139 self.SetDefaultGateway(gateway) self.SetDefaultGateway(gateway)
6140 name = "DefaultGateway" name = "DefaultGateway"

If you observe line numbers from 1701-1703, that seems like a duplicate entry
in the file “/usr/sbin/waagent(from WALinuxAgent-2.0.18-1.noarch)” which could
be probably the reason why we are getting two entries for default gateway and
necessarily with same IP’s.
```

**Comment by hglkrijger** [comment]
Isn't this a duplicate of #190?

**Comment by hglkrijger** [comment]
Not sure who originally reported this issue, but we would need more information to go any further. From the information above, this looks like a duplicate which has already been fixed and will be available with 2.1.5+ of the agent. I am closing for now, but if this can be reproduced on the latest agent we can take another look.
